### DISCUSSION TITLE
No mod yet, but lost telnet

### DISCUSSION THREAD

**Discussion by b-fitzpatrick** [xblax/flashforge_ad5m_klipper_mod#157] [General]
I know this isn't related to this project, but I'm hoping someone here may have some insight. I will probably end up installing this mod, but I first wanted to see if I could get by with just some minor tweaks using root access. I started down that path using the process here: [Root access](https://github.com/g992/flashforge-ad5m-5mpro-research/blob/main/upgrades/getting_root.md) However, after the first step, I could not telnet into the printer. I get "Connection refused". I can ping it.

Some things that may or may not be related:
- I updated to 2.6.5 OTA when I first got it.
- I was able to connect via telnet at some point. I made a couple stupid password attempts and moved on.
- I downgraded to 2.4.5 as part of the root process. I thought a full firmware "flash" was required to run the script, and 2.6.5 isn't available for download. I've since learned that I can just run whatever I want in the script.
- During the first step of the root process, I included the 'adduser "$USERNAME" sudo' that's shown in the screenshot. I don't think this did anything, as there's no 'sudo' group.
- I enabled the new "LAN mode" in 2.6.5. I think it was enabled when I downgraded to 2.4.5. I reinstalled 2.6.5 to disable it in case it's somehow related.
- There are what appear to be backups of /etc/passwd, /etc/group, and /etc/shadow. They don't have the user that I added in the first step of attempting root access. They have a dash appended to the filename. I found these by copying /etc/* to the USB drive. 

Any ideas to restore telnet access?

**Comment by consp** [comment]
> I enabled the new "LAN mode" in 2.6.5. I think it was enabled when I downgraded to 2.4.5. 

Looking at the scripts for the OTA update, I'm pretty sure they are persistent after downgrade. There are some network changes in there, haven't looked too much into them to check if they alter telnet or eth0

> I could not telnet into the printer. I get "Connection refused". I can ping it.
> Any ideas to restore telnet access?

What does `nmap -sT -T4 [printer ip]` say?

Easiest would be to install the mod, ssh to the machine, goto `/mnt/orig_mnt/` (dir name might be a bit different but you get the gist, it's the original filesystem). Goto /mnt/orig_mnt/etc and check if telnet is still available in init.d and enabled or check the `/mnt/orig_mnt/var/log` dir for errors. Might also want to do `mount -o remount,rw /mnt/orig_mnt` to get write access if needed.

YMMV applies when messing with the original filesystem, especially in write mode.

**Reply by b-fitzpatrick** [reply]
Thank you!

nmap shows only two ports open:
```
PORT     STATE SERVICE
8080/tcp open  http-proxy
8899/tcp open  ospf-lite
```

Saved the output of "ls -Flah /etc/init.d" to the USB drive with the boot script, and telnet seems to be there and executable.
```
total 52
drwxr-xr-x    2 root     root        4.0K Jan  1 00:00 ./
drwxr-xr-x   16 root     root        4.0K Jan  1 00:00 ../
-rwxr-xr-x    1 root     root        1.0K Aug 25  2023 S01syslogd*
-rwxr-xr-x    1 root     root        1.0K Aug 25  2023 S02klogd*
-rwxr-xr-x    1 root     root        1.6K Aug 25  2023 S10udev*
-rwxr-xr-x    1 root     root        1.2K Aug 25  2023 S20urandom*
-rwxr-xr-x    1 root     root        1.6K Aug 25  2023 S30dbus*
-rwxr-xr-x    1 root     root         438 Aug 25  2023 S40network*
-rwxr-xr-x    1 root     root         642 Aug 25  2023 S50postgresql*
-rwxr-xr-x    1 root     root         614 Jul 28  2023 S50telnet*
-rwxr-xr-x    1 root     root         427 Aug 25  2023 S80dnsmasq*
-rwxr-xr-x    1 root     root         423 Aug 25  2023 rcK*
-rwxr-xr-x    1 root     root         960 Jan  1 00:00 rcS*
```

Will have to keep digging.

**Reply by consp** [reply]
> telnet seems to be there and executable

Doesn't mean it is starting though, might be unbound to the interfaces. as well. You tried both ethernet and wlan? 

> 8080/tcp open  http-proxy
> 8899/tcp open  ospf-lite

Seems like only the webcam and the flashforge protocol are exposed. 

I can now confirm telnet is disabled in 2.6.5, I'll have a look as to why.


Sidenote/Funny thing: flashforge plays the startup song twice and the second time overwrites the first one. So in case you were wondering why it sounds weird in 2.6.5 that's the reason.

**Reply by consp** [reply]
Well that's nasty:

`system("mv /usr/sbin/telnetd /usr/sbin/myBin");`

Is triggered when `/usr/sbin/telnetd` is found by the 2.6.5 `firmwareExe` file. You'll have to alter telnetd in S50telnet to either `/bin/busybox` (it's a symlink after all) or the new file `/usr/sbin/myBin`. Maybe do some `sed -i` magic to get it working again.

I guess they didn't like people getting into the system ... there is a serial console installed so that's pretty pointless anyways ugh.

It's triggered just before the startup song so if you hear that it's already done.


Contents of S50telnet:

```
#!/bin/sh
#
# Start telnet....
#

TELNETD_ARGS=-F
[ -r /etc/default/telnet ] && . /etc/default/telnet

start() {
      printf "Starting telnetd: "
      start-stop-daemon -S -q -m -b -p /var/run/telnetd.pid \
                        -x /usr/sbin/telnetd -- $TELNETD_ARGS
      [ $? = 0 ] && echo "OK" || echo "FAIL"
}

stop() {
        printf "Stopping telnetd: "
        start-stop-daemon -K -q -p /var/run/telnetd.pid \
                          -x /usr/sbin/telnetd
        [ $? = 0 ] && echo "OK" || echo "FAIL"
}

case "$1" in
    start)
        start
        ;;
    stop)
        stop
        ;;
    restart|reload)
        stop
        start
        ;;
  *)
        echo "Usage: $0 {start|stop|restart}"
        exit 1
esac
```

My guess adding `sed -i 's/sbin\/telnetd/bin\/busybox/g /etc/init.d/S50telnet` to the startup script and then only running that via usb might do the trick but this is definetly not recommended and at your own risk.

**Reply by b-fitzpatrick** [reply]
Thanks for figuring it out. Bummer that they'd do that. It worries me a bit for future of this project. Should it?

**Reply by consp** [reply]
> It worries me a bit for future of this project. Should it?

Since the only update mechanism is via a script ran as root I don't think there will be a problem.

**Reply by b-fitzpatrick** [reply]
I've attempted to downgrade back to 2.4.5 to install the mod, but it's not working. I put the files on the USB drive like I did before. I get an indication that it's upgrading and then instructions to remove the USB drive and reboot. It only takes a few seconds between these two screens. Then, I reboot, and it boots normally into 2.6.5.

Did the contents of auto_run.sh change? I don't see anything that would prevent the downgrade it, but I could be missing something. I am able to copy files to the USB drive and poke around, but it's cumbersome without terminal access.

```
#!/bin/sh

WORK_DIR=/opt
MACHINE=Adventurer5MPro
PID=0024

#cat /opt/boot.img > /dev/fb0


if [ ! "`mount | grep "mmcblk0p7"`" ]; then
	echo "mmcblk0p7 not mounted and will fsck";
	fsck -y /dev/mmcblk0p7 && mount /dev/mmcblk0p7 /data;
fi


if [ -d /data/update ];then
	rm -rf /data/update
fi

insmod /lib/modules/8821cu.ko
#wait for USB
sleep 2

for i in 1 2 3 4;
do
  if [ ! -e /dev/sda$i ]; then
     echo "sda$i not exist"
	 if [ ! -e /dev/sda ];then
     	continue
	 else
	 	echo "find /dev/sda. start mount."
  		mount -t vfat -o,codepage=936,iocharset=utf8 /dev/sda /mnt
	 fi
  else
  	mount -t vfat -o,codepage=936,iocharset=utf8 /dev/sda$i /mnt
  fi

  if [ $? -ne 0 ]; then
        echo "mount /dev/sda or /dev/sda$i to /mnt failed"
        continue
  else
  		ls -1t /mnt/Adventurer5MPro-*.tgz
		if [ $? -eq 0 ];then
			UPDATEFILE=`ls -1t /mnt/Adventurer5MPro-*.tgz | head -n 1`
			if [ -f $UPDATEFILE ];then
				echo "find update file: ${UPDATEFILE}"
				rm -rf /data/update
				cp -a ${UPDATEFILE} /data/
				if [ $? -ne 0 ];then
					rm -rf /data/Adventurer5MPro-*.tgz
					sync
					umount /mnt
					break
				fi
				sync
				mkdir -p /data/update
				sync
				SRCFILE="/data/`basename ${UPDATEFILE}`"
				if [ -f ${SRCFILE} ];then
					tar -xvf ${SRCFILE} -C /data/update/
					sync
					rm -rf ${SRCFILE}
					/data/update/flashforge_init.sh ${MACHINE} ${PID}
					if [ $? -eq 0 ];then
						umount /mnt
						rm -rf /data/update
						sleep 100000
					fi
					umount /mnt
					rm -rf /data/update
					break
				fi
			fi
		fi

        if [ -f /mnt/flashforge_init.sh ]; then
             echo "found /mnt/flashforge_init.sh"
             chmod a+x /mnt/flashforge_init.sh
             /mnt/flashforge_init.sh ${MACHINE} ${PID}
			 if [ $? -eq 0 ];then
				umount /mnt
				sleep 100000
			 fi
             umount /mnt
             break
        fi
        umount /mnt
  fi
done

/opt/klipper/usb_map.sh

CONTROL_DIR=/opt/PROGRAM/control/
CONTROL_VERSION=`ls -1t ${CONTROL_DIR} | head -n 1`
CONTRIL_FLAG=${CONTROL_DIR}${CONTROL_VERSION}/Update
echo "check update control file"
echo $CONTRIL_FLAG

if [ -f ${CONTRIL_FLAG} ];then
	cd ${CONTROL_DIR}${CONTROL_VERSION}
	./run.sh

	reboot -f
fi

rm /opt/config/printer-*
rm /data/logFiles/printer.log*

/opt/PROGRAM/ffstartup-arm -f /opt/PROGRAM/ffstartup.cfg &
```

**Reply by consp** [reply]
> I've attempted to downgrade back to 2.4.5 to install the mod, but it's not working.

I think it would be easier to change the install script of the mod (device_files/flashforge_init.sh) and remove the following lines:
```
    FF_VERSION="$(cat /root/version)"
    if [ "$FF_VERSION" != "2.4.5" ]
    then
        echo "Printer software version not supported."
        xzcat $WORK_DIR/img/install_fail_vers.img.xz > /dev/fb0
        exit 0
    fi
```

And replace the one in the mod tarball (or edit that one specifically). Note: the tgz files are *not* gzipped despite their name.
Easiest would be to just copy all contents to a usb drive (extract), you do not need the tarball at all. That way you also ignore the autorun.sh. I'm pretty sure the original flashforge_init.sh will also check for one single version so "downgrade" is not really a thing that way (have not confirmed!).

Note: while not recommended yet (simply because it's in testing), I've had no problems so far installing the mod on 2.6.5 since no system files have changed so altering the flashforge_init.sh script should be fine.

Do make sure you do _not_ do `rm -rf /` or something similar in the flashforge_init.sh script ;)

**Reply by b-fitzpatrick** [reply]
I deleted a comment about the root partition being full. Turns out that was my mistake. Once when I was dumping files to the USB drive to poke around, I forgot "$WORK_DIR", so it was copying files to a folder in the root partition.

**Reply by consp** [reply]
> Once when I was dumping files to the USB drive to poke around, I forgot "$WORK_DIR", so it was copying files to a folder in the root partition.

Would have been surprized. Afaik it's a 8GB eMMC chip so there should be plenty of space.

**Reply by b-fitzpatrick** [reply]
The root partition is limited to 1.2G:
```
Filesystem                Size      Used Available Use% Mounted on
/dev/root                 1.2G    760.3M    463.6M  62% /
tmpfs                    53.9M     28.0K     53.9M   0% /tmp
tmpfs                    53.9M     28.0K     53.9M   0% /run
devtmpfs                 44.7M         0     44.7M   0% /dev
/dev/mmcblk0p7            5.8G    479.3M      5.0G   9% /data
/dev/sda1                14.3G    969.2M     13.4G   7% /mnt
```
This is AFTER I fixed it. Mod installed and working well, now. Thanks to all who have contributed!

**Reply by consp** [reply]
> The root partition is limited to 1.2G:

Ah, forgot the data partition is separate ... make sense now. Good call.

> Mod installed and working well, now. 

You can fix the telnet now on the original firmware :)
